src/main/java/net/minecraftforge/common/capabilities/CapabilityDispatcher.java

```diff
@@ -5,139 +5,138 @@
-import net.minecraftforge.common.util.LazyOptional;
-import org.jetbrains.annotations.Nullable;
 
 /**
- * A high-speed implementation of a capability delegator.
- * This is used to wrap the results of the AttachCapabilitiesEvent.
- * It is HIGHLY recommended that you DO NOT use this approach unless
- * you MUST delegate to multiple providers instead just implement y
- * our handlers using normal if statements.
- *
- * Internally the handlers are baked into arrays for fast iteration.
- * The ResourceLocations will be used for the NBT Key when serializing.
+ * A Capability Dispatcher is a manager object that contains the result of the
+ * {@link AttachCapabilitiesEvent} and makes the attached capabilities visible on
+ * the parent object.
+ * 
+ * @param <T> The type of the parent object.
  */
 @ParametersAreNonnullByDefault
 @MethodsReturnNonnullByDefault
-public final class CapabilityDispatcher implements INBTSerializable<CompoundTag>, ICapabilityProvider
+public final class CapabilityDispatcher<T extends ICapabilityProvider> implements INBTSerializable<CompoundTag>, ICapabilityProvider
 {
-    private ICapabilityProvider[] caps;
-    private INBTSerializable<Tag>[] writers;
-    private String[] names;
-    private final List<Runnable> listeners;
 
-    public CapabilityDispatcher(Map<ResourceLocation, ICapabilityProvider> list, List<Runnable> listeners)
+    private final T owner;
+    private final Map<CapabilityType<?>, IAttachedCapabilityProvider<?, T>> caps;
+    private final Map<ResourceLocation, IAttachedCapabilityProvider<?, T>> byName;
+    private boolean isValid = true;
+
+    public CapabilityDispatcher(AttachCapabilitiesEvent<T> event, T owner)
     {
-        this(list, listeners, null);
+        this.caps = event.getCapabilities();
+        this.byName = event.getCapabilitiesByName();
+        this.owner = owner;
     }
 
-    @SuppressWarnings("unchecked")
-    public CapabilityDispatcher(Map<ResourceLocation, ICapabilityProvider> list, List<Runnable> listeners, @Nullable ICapabilityProvider parent)
+    /**
+     * Internal copy constructor.
+     * @see {@link #copy(ICapabilityProvider)}
+     * @see {@link IAttachedCapabilityProvider#copy(ICapabilityProvider)}
+     */
+    private CapabilityDispatcher(CapabilityDispatcher<T> other, T newOwner)
     {
-        List<ICapabilityProvider> lstCaps = Lists.newArrayList();
-        List<INBTSerializable<Tag>> lstWriters = Lists.newArrayList();
-        List<String> lstNames = Lists.newArrayList();
-        this.listeners = listeners;
-
-        if (parent != null) // Parents go first!
-        {
-            lstCaps.add(parent);
-            if (parent instanceof INBTSerializable)
-            {
-                lstWriters.add((INBTSerializable<Tag>)parent);
-                lstNames.add("Parent");
-            }
-        }
+        this.caps = other.caps.values().stream().map(old -> old.copy(newOwner)).collect(Collectors.toMap(IAttachedCapabilityProvider::getType, t -> t));
```

Comment:
Please avoid any kind of stream, this code path is super hot, so make it as fast and explicit as possible.